tests/main/apparmor-prompting-integration-tests/task.yaml

```diff
@@ -0,0 +1,110 @@
+  - ubuntu-2*
+  - ubuntu-core-*
+
+environment:
+    VARIANT/read_single_allow: read_single_allow
+    VARIANT/read_single_deny: read_single_deny
+    VARIANT/write_single_allow: write_single_allow
+    VARIANT/write_single_deny: write_single_deny
+    VARIANT/create_multiple_allow: create_multiple_allow
+    VARIANT/create_multiple_deny: create_multiple_deny
+    VARIANT/download_file_defaults: download_file_defaults
+    VARIANT/download_file_safer: download_file_safer
+    VARIANT/download_file_conflict: download_file_conflict
+    VARIANT/timespan_allow: timespan_allow
+    VARIANT/timespan_deny: timespan_deny
+    VARIANT/create_multiple_actioned_by_other_pid_always_allow: create_multiple_actioned_by_other_pid_always_allow
+    VARIANT/create_multiple_actioned_by_other_pid_always_deny: create_multiple_actioned_by_other_pid_always_deny
+    VARIANT/create_multiple_not_actioned_by_other_pid_single_allow: create_multiple_not_actioned_by_other_pid_single_allow
+    VARIANT/create_multiple_not_actioned_by_other_pid_single_deny: create_multiple_not_actioned_by_other_pid_single_deny
+    VARIANT/write_read_multiple_actioned_by_other_pid_allow_deny: write_read_multiple_actioned_by_other_pid_allow_deny
+    VARIANT/write_read_multiple_actioned_by_other_pid_deny_allow: write_read_multiple_actioned_by_other_pid_deny_allow
+
+prepare: |
+    tests.session prepare -u test
+    #shellcheck disable=SC2016
+    tests.session -u test exec sh -c 'mkdir -p "$HOME/integration-tests"' # $HOME must be that of test user
+    snap install prompting-client
+
+restore: |
+    #shellcheck disable=SC2016
+    tests.session -u test exec sh -c 'rm -rf "$HOME/integration-tests"' # $HOME must be that of test user
+    tests.session restore -u test
+
+debug: |
+    TEST_UID="$(id -u test)"
+    echo "outstanding prompts:"
+    snap debug api "/v2/interfaces/requests/prompts?user-id=$TEST_UID"
+    echo "rules:"
+    snap debug api "/v2/interfaces/requests/rules?user-id=$TEST_UID"
+
+execute: |
+    echo "Enable prompting via snap client where possible"
+    # Prompting is unsupported everywhere but the Ubuntu non-core systems with
+    # kernels which support apparmor prompting
+    if ! os.query is-ubuntu || os.query is-core || ! grep 'prompt' /sys/kernel/security/apparmor/features/policy/permstable32 ; then
+        not snap set system experimental.apparmor-prompting=true >& err.out
+        if os.query is-core ; then
+            MATCH "cannot enable prompting feature as it is not supported on Ubuntu Core systems" < err.out
+        else
+            MATCH "cannot enable prompting feature as it is not supported by the system" < err.out
+        fi
+
+        # even if unsupported setting it to false should succeed
+        snap set system experimental.apparmor-prompting=false
+        exit 0
+    fi
+
+    echo "Enable AppArmor prompting experimental feature"
+    snap set system experimental.apparmor-prompting=true
+
+    # Wait for snapd to begin restart
+    sleep 1
```

Comment:
the retry is not enough?

Reply:
My concern was if the retry succeeded before the restart even began. But if we know the restart begins before the API request completes, we can rely on the retry. I'm sure if this is the case...

Reply:
you could look at journalctl -u snapd  to know if a restart has happend I suppose, we might even already some helper to check things in logs?

Reply:
I used the MainPID check to ensure the PID has changed:
```diff
iff --git a/tests/main/apparmor-prompting-integration-tests/task.yaml b/tests/main/apparmor-prompting-integration-tests/task.yaml
index 64a7adaaef..f05e993fae 100644
--- a/tests/main/apparmor-prompting-integration-tests/task.yaml
+++ b/tests/main/apparmor-prompting-integration-tests/task.yaml
@@ -55,6 +55,10 @@ debug: |
     snap debug api "/v2/interfaces/requests/rules?user-id=$TEST_UID" || true
 
 execute: |
+    echo "Precondition check that snapd is active"
+    systemctl is-active snapd.service snapd.socket
+    SNAPD_PID="$(systemctl show --property MainPID snapd.service | cut -f2 -d=)"
+
     echo "Enable prompting via snap client where possible"
     # Prompting is unsupported everywhere but the Ubuntu non-core systems with
     # kernels which support apparmor prompting
@@ -72,10 +76,10 @@ execute: |
     echo "Enable AppArmor prompting experimental feature"
     snap set system experimental.apparmor-prompting=true
 
-    # Wait for snapd to begin restart
-    sleep 1
+    echo "Wait for snapd to begin restart"
+    retry --wait 1 -n 60 test "$SNAPD_PID" != "$(systemctl show --property MainPID snapd.service | cut -f2 -d=)"
 
-    echo "Wait until prompting is active"
+    echo "Wait until snapd is active"
     retry --wait 1 -n 60 systemctl is-active snapd
 
     echo "Check that shell script and scripted replies exist for $VARIANT"
```